native/core/src/activities.rs

```diff
@@ -1,55 +1,624 @@
+            ActivityContent::TaskDueDateChange {
+                new_due_date: Some(new_due_date),
+                ..
+            } => Some(UtcDateTime::from_naive_utc_and_offset(
+                new_due_date.and_time(
+                    NaiveTime::from_num_seconds_from_midnight_opt(0, 0).expect("midnight exists"),
+                ),
+                Utc,
+            )),
+            _ => None,
+        }
+    }
+
+    pub fn ref_details(&self) -> Option<RefDetails> {
+        if let ActivityContent::Reference { details, .. } = &self.inner {
+            Some(details.clone())
+        } else {
+            None
+        }
+    }
+
+    pub fn target_url(&self) -> String {
+        match &self.inner {
+            ActivityContent::Boost { .. } => format!("/updates/{}", self.meta.event_id),
+            ActivityContent::TitleChange { object, .. }
+            | ActivityContent::DescriptionChange { object, .. }
+            | ActivityContent::EventDateChange { object, .. }
+            | ActivityContent::Rsvp { object, .. }
+            | ActivityContent::TaskProgress { object, .. }
+            | ActivityContent::TaskDueDateChange { object, .. }
+            | ActivityContent::TaskAccept { object, .. }
+            | ActivityContent::TaskDecline { object, .. }
+            | ActivityContent::OtherChanges { object }
+            | ActivityContent::Creation { object, .. } => object.target_url(),
 
-    fn try_from(mdl: AnyActerModel) -> Result<Self, Self::Error> {
-        if let AnyActerModel::RoomStatus(s) = mdl {
-            match s.inner {
+            ActivityContent::Attachment { object, .. } => format!(
+                "{}?section=attachments&attachmentId={}",
+                object.target_url(),
+                encode(self.meta.event_id.as_str()),
+            ),
+            ActivityContent::Reference { object, .. } => format!(
+                "{}?section=references&referenceId={}",
+                object.target_url(),
+                encode(self.meta.event_id.as_str()),
+            ),
+            ActivityContent::Comment { object, .. } => format!(
+                "{}?section=comments&commentId={}",
+                object.target_url(),
+                encode(self.meta.event_id.as_str()),
+            ),
+            ActivityContent::Reaction { object, .. } => format!(
+                "{}?section=reactions&reactionId={}",
+                object.target_url(),
+                encode(self.meta.event_id.as_str()),
+            ),
+
+            ActivityContent::TaskAdd { object, .. } => {
+                format!("/tasks/{}/{}", object.object_id_str(), self.meta.event_id)
+            }
+            ActivityContent::MembershipChange(_)
+            | ActivityContent::RoomCreate(_)
+            | ActivityContent::RoomName(_) => todo!(),
+        }
+    }
+}
+
+impl Activity {
+    pub async fn for_acter_model(store: &Store, mdl: AnyActerModel) -> Result<Self, crate::Error> {
```

Comment:
we are now converting the activity right from the model, but the logic is basically the same as before and most of this is still covered by the same and additional tests.